src/MLIR/Native.hs

```diff
@@ -303,6 +348,16 @@ showSomething action = do
       free dataPtr
       return bs
 
+-- | Unroll using a function that is equivalent to "get next" inside IO.
+unrollIOMaybe :: (a -> IO (Maybe a)) -> IO (Maybe a) -> IO [a]
+unrollIOMaybe fn z = do
+  x <- z
+  case x of
+      Nothing -> return mzero
+      Just x' -> do
+          xs <- unrollIOMaybe fn (fn x')
+          return (return x' `mplus` xs)
```

Comment:
Perhaps just `return $ x' : xs`?

Reply:
And if you want to be very fancy then you can fold this whole `do` block into `(x':) <$> unrollIOMaybe fn (fn x')`